src/lib/utils/bitvector.h

```diff
@@ -0,0 +1,1327 @@
+            bitref(const bitref&) noexcept = default;
+            bitref(bitref&&) noexcept = default;
+
+            constexpr bitref& set() noexcept {
+               this->m_block |= this->m_mask;
+               return *this;
+            }
+
+            constexpr bitref& unset() noexcept {
+               this->m_block &= ~this->m_mask;
+               return *this;
+            }
+
+            constexpr bitref& flip() noexcept {
+               this->m_block ^= this->m_mask;
+               return *this;
+            }
+
+            // NOLINTBEGIN
+
+            constexpr bitref& operator=(bool bit) noexcept {
+               this->m_block =
+                  CT::Mask<BlockT>::expand(bit).select(this->m_mask | this->m_block, this->m_block & ~this->m_mask);
+               return *this;
+            }
+
+            constexpr bitref& operator=(const bitref& bit) noexcept { return *this = bit.is_set(); }
+
+            constexpr bitref& operator=(bitref&& bit) noexcept { return *this = bit.is_set(); }
+
+            // NOLINTEND
+
+            constexpr bitref& operator&=(bool other) noexcept {
+               this->m_block &= ~CT::Mask<BlockT>::expand(other).if_not_set_return(this->m_mask);
+               return *this;
+            }
+
+            constexpr bitref& operator|=(bool other) noexcept {
+               this->m_block |= CT::Mask<BlockT>::expand(other).if_set_return(this->m_mask);
+               return *this;
+            }
+
+            constexpr bitref& operator^=(bool other) noexcept {
+               this->m_block ^= CT::Mask<BlockT>::expand(other).if_set_return(this->m_mask);
+               return *this;
+            }
+      };
+
+   public:
+      bitvector_base() : m_bits(0) {}
+
+      bitvector_base(size_type bits) : m_bits(bits), m_blocks(ceil_toblocks(bits)) {}
+
+      /**
+       * Initialize the bitvector from a byte-array. Bits are taken byte-wise
+       * from least significant to most significant. Example::
+       *
+       *    bitvector[0] -> LSB(Byte[0])
+       *    bitvector[1] -> LSB+1(Byte[0])
+       *    ...
+       *    bitvector[8] -> LSB(Byte[1])
+       *
+       * @param bytes The byte vector to be loaded
+       * @param bits  The number of bits to be loaded. This must not be more
+       *              than the number of bytes in @p bytes.
+       */
+      bitvector_base(std::span<const uint8_t> bytes, std::optional<size_type> bits = std::nullopt) {
+         from_bytes(bytes, bits);
+      }
+
+      bitvector_base(std::initializer_list<block_type> blocks, std::optional<size_type> bits = std::nullopt) :
+            m_bits(bits.value_or(blocks.size() * block_size_bits)), m_blocks(blocks.begin(), blocks.end()) {}
+
+      bool empty() const { return m_bits == 0; }
+
+      size_type size() const { return m_bits; }
+
+      /**
+       * @returns true iff the number of 1-bits in this is odd, false otherwise
+       */
+      bool ct_has_odd_hamming_weight() const {
```

Comment:
To avoid issues with range analysis it might be useful to return `CT::Choice` here